src/nf/nf_rnn_layer_submodule.f90

```diff
@@ -0,0 +1,169 @@
+
+  end subroutine forward
+
+
+  pure module function get_num_params(self) result(num_params)
+    class(rnn_layer), intent(in) :: self
+    integer :: num_params
+
+    ! Number of weigths times number of biases
+    num_params = self % input_size * self % output_size &
+                 + self % output_size * self % output_size &
+                 + self % output_size
+
+  end function get_num_params
+
+
+  pure module function get_params(self) result(params)
+    class(rnn_layer), intent(in) :: self
+    real, allocatable :: params(:)
+
+    params = [ &
+      pack(self % weights, .true.), &
+      pack(self % recurrent, .true.), &
+      pack(self % biases, .true.) &
+    ]
+
+  end function get_params
+
+
+  pure module function get_gradients(self) result(gradients)
+    class(rnn_layer), intent(in) :: self
+    real, allocatable :: gradients(:)
+
+    gradients = [ &
+      pack(self % dw, .true.), &
+      pack(self % db, .true.) &
+    ]
+
+  end function get_gradients
+
+
+  module subroutine set_params(self, params)
+    class(rnn_layer), intent(in out) :: self
+    real, intent(in) :: params(:)
+    integer :: first, last
+
+    ! check if the number of parameters is correct
+    if (size(params) /= self % get_num_params()) then
+      error stop 'Error: number of parameters does not match'
+    end if
+
+    ! reshape the weights
+    last = self % input_size * self % output_size
+    self % weights = reshape( &
+      params(:last), &
+      [self % input_size, self % output_size] &
+    )
+
+    ! reshape the recurrent weights
+    first = last + 1
+    last = first + self % output_size * self % output_size
+    self % recurrent = reshape( &
+      params(first:last), &
+      [self % output_size, self % output_size] &
+    )
+
+    ! reshape the biases
+    first = last + 1
+    last = first + self % output_size
+    self % biases = reshape( &
+      params(first:last), &
+      [self % output_size] &
+    )
+
+  end subroutine set_params
+
+
+  module subroutine init(self, input_shape)
+    class(rnn_layer), intent(in out) :: self
+    integer, intent(in) :: input_shape(:)
+
+    self % input_size = input_shape(1)
+
+    ! Weights are a 2-d array of shape previous layer size
+    ! times this layer size.
+    allocate(self % weights(self % input_size, self % output_size))
+    call random_normal(self % weights)
+    self % weights = self % weights / self % input_size
+
+    ! Broadcast weights to all other images, if any.
+    call co_broadcast(self % weights, 1)
+
+    allocate(self % recurrent(self % output_size, self % output_size))
+    call random_normal(self % recurrent)
+    self % recurrent = self % recurrent / self % input_size
+
+
+    allocate(self % biases(self % output_size))
+    self % biases = 0
+
+    allocate(self % output(self % output_size))
+    self % output = 0
+
+    allocate(self % z(self % output_size))
+    self % z = 0
+
+    allocate(self % state(self % output_size))
+    self % state = 0
```

Comment:
@milancurcic , I have a question for you. I'm implementing just a basic RNN with no warm-up. For every new sequence, this state should be reset to some initial value, let's say zero. There are two options here:
- The inputs could be a 2D array, sequence x feature. I tried to implement it, but I think I was breaking logical coherence with the 3D family of procedures.
- Some 'reset()' procedure to avoid needing to re-instantiate the net for every new sequence.

What do you suggest?

A note, I just realized that predict() is based on forward(), thus it necessarily update the state while predicting.

Reply:
State is what gives the RNN layer memory, correct? So, we want to reset the state between each sequence in both training and inference, right? Calling `init` doesn't work because we only want to reset `state`, but not the learned parameters of the layer. If I got this correctly and you agree, then I think a `reset()` subroutine that is specific to the RNN layer makes perfect sense.

For the first question

> The inputs could be a 2D array, sequence x feature. I tried to implement it, but I think I was breaking logical coherence with the 3D family of procedures.

It can be done (and we should, eventually), but for simplicity and to keep this PR manageable, let's for now stick with the simple case of 1-d input (not sure if that's one value of each feature or a sequence of one feature; you let me know, I've never used or studied much the recurrent layers).

> A note, I just realized that predict() is based on forward(), thus it necessarily update the state while predicting.

Yes, `predict` is just a convenience function wrapper around the `forward` subroutine to return the output of the network.

Reply:
Yes and Yes.

Sure, `reset()` sounds like a good way. Good point on keeping the PR small. I think it would be more powerful to allow multiple features, so this 1D would be features, and the sequence evolution can be obtained with multiple calls on `forward()` or `predict()` since the state is preserved.

Thanks!

Reply:
@milancurcic , could you take a look, please, in the new `reset()`? It wasn't clear to me what defines a pure/impure module.

I didn't realized at first, but it should be accessible in the top net level since if can be composed with multiple RNN layers.

I haven't included in the PR yet, but it pass my local tests.

Reply:
On second thought, do you think the resetting of the RNN layer state can be done automatically by the library code, or does it have to be done by the user? It seems to me a but unwieldy to have `nework % reset()` that's specific to recurrent layers only. I see that Keras allows `initial_state` to be passed to the layer (and initializes to zero by default): https://keras.io/api/layers/recurrent_layers/simple_rnn/.

Do you think it would be a good approach to define the following subroutines:

in `nf_layer.f90`

```fortran
subroutine set_state(self, state)
  type(layer), intent(inout) :: self
  real, intent(in), optional :: state(:)
  if (present(state)) then
    ! set the state on the rnn_layer concrete instance
  else
    ! set the state to zero on the rnn_layer concrete instance
  end if
  ...
end subroutine set_state
```

and in `nf_rnn_layer.f90`:

```fortran
subroutine set_state(self, state)
  type(rnn_layer), intent(inout) :: self
  real, intent(in), optional :: state(:)
  if (present(state)) then
    self % state = state
  else
    self % state = 0
  end if
end subroutine set_state
```

But not defining it on the network level, but rather setting it directly on the layers (this would be in user's application code):

```fortran
! Reset the state on all recurrent layers
do n = 1, size(net % layers)
  if (net % layers(n) % name == 'rnn') call net % layers(n) % set_state()
end do
```

Reply:
I see your point. Adding a reset in the top level that would only be useful for RNNs is something.

Yes, that should work. Eventually the N-D array input where we can define, or have a convention on, which dimension is the sequence would be the best since it could reset itself.

I'll update the PR with your suggestion. Thanks for your time.

Reply:
@milancurcic , is this close to what you had in mind?

Reply:
Yes, excellent! Do you think this is ready for me to start playing with locally?

Reply:
It is missing a few things, including better tests. I also think that I messed up in the back propagation of the error. I'll check that on the first chance. I've been calibrating my NN outside, with TensorFlow itself.

But, if you have the time, it would be great if you could check it sooner than later for logical errors. I don't know modern fortran.